tools/generators/provider-generator/index.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  formatFiles,
+  generateFiles,
+  getWorkspaceLayout,
+  joinPathFragments,
+  logger,
+  names,
+  readJson,
+  readProjectConfiguration,
+  Tree,
+  updateJson,
+  updateProjectConfiguration,
+} from '@nrwl/devkit';
+import { libraryGenerator } from '@nrwl/node';
+import { join } from 'path';
+
+interface NewProviderSchemaOptions {
+  name: string;
+}
+
+export default async function (tree: Tree, schema: NewProviderSchemaOptions) {
+  const { className, fileName } = names(schema.name);
+  const libClassName = `${className}Provider`;
+  const targetName = `${fileName}-demo`;
+  const libName = `js-${fileName}-provider`;
+  const { libsDir } = getWorkspaceLayout(tree);
+  const projectRoot = joinPathFragments(libsDir, libName);
+  const projectLibDir = joinPathFragments(projectRoot, 'src', 'lib');
+
+  /**
+   * Use the OOTB NX TypeScript library generator
+   */
+  await libraryGenerator(tree, {
+    name: libName,
+    compiler: 'tsc',
+    buildable: true,
+    js: false,
+    strict: true,
+  });
+
+  /**
+   * Change the build mode to webpack
+   */
+  const currentConfig = readProjectConfiguration(tree, libName);
+  if (currentConfig.targets) {
+    currentConfig.targets.build = {
+      executor: '@nrwl/node:webpack',
+      outputs: ['{options.outputPath}'],
+      options: {
+        outputPath: `dist/packages/${libName}`,
+        main: `packages/${libName}/src/index.ts`,
+        tsConfig: `packages/${libName}/tsconfig.lib.json`,
+      },
+      configurations: {
+        production: {
+          optimization: true,
+          extractLicenses: true,
+          inspect: false,
+        },
+      },
+    };
+  }
+  updateProjectConfiguration(tree, libName, currentConfig);
+
+  /**
+   * Refactors the files
+   */
+  ['spec.ts', 'ts'].forEach((suffix) => {
+    tree.rename(
+      joinPathFragments(projectLibDir, `${libName}.${suffix}`),
+      joinPathFragments(projectLibDir, `${fileName}.${suffix}`)
+    );
+  });
+
+  /**
+   * Adding provider code
+   */
+  generateFiles(tree, join(__dirname, 'files', 'provider'), projectRoot, {
+    name: schema.name,
+    fileName,
+    libClassName,
+    tmpl: '',
+  });
+
+  /**
+   * Adding demo script
+   */
+  generateFiles(tree, join(__dirname, 'files', 'script'), 'scripts', {
+    name: schema.name,
+    fileName,
+    libClassName,
+    libName,
+    tmpl: '',
```

Comment:
is this empty string needed? We cant omit it?

Reply:
It's used in the file path. I think it's just used to verify the template strings are working. NX used this pattern in their examples.